### PR TITLE
New user walkthrough. Closes #1895

### LINKED ISSUE #1895: Create detailed walkthrough for new users
With the expectation of new users wholly unaffiliated with the DeepForge team, it would be best for the docs to include a detailed walkthrough on how to create a functional pipeline from scratch. In line with recent work to integrate with SciServer, this walkthrough should use motivating examples relevant to natural science users, such as the already established redshift estimation task.

### PR DESCRIPTION
**Pull request by KyleAMoore**
A full walkthrough has been created for new users to learn the basics of creating a functional project from a blank starting point. The walkthrough has numerous sections that detail how to create a pipeline and its constituent parts. There are also two step-by-step guides on creating a simple and advanced pipeline from scratch.

### CODE REVIEW

**Comment by KyleAMoore** [comment]
It seems prudent to export and upload the project created in the step-by-step tutorials. It is not clear, however, whether the exported project should be included in the walkthrough subdirectory within the docs directory or if it should be added as another example in https://github.com/deepforge-dev/examples.

Either way, a link to the exported project should be added to the walkthrough before PR is accepted if the project is to be published.

**Comment by KyleAMoore** [comment]
Navigation links added to all pages except `walkthrough.rst`. Each page has a link to the previous and next tutorials, as well as a link back to the index.

I believe that these links should be and are centered on the page. They do not render centered on GitHub, but they have been given the class `.centered`, which exists in the https://deepforge.readthedocs.io css (at the 19th and 20th occurrence of the string "center" on [this page](https://deepforge.readthedocs.io/en/latest/_static/css/theme.css)).

I would prefer it if there were more space between the three links for ease use and distinguishability. Ideally, the links would each be aligned separately so that the next and previous links are in the corners and only the index links are centered. I was unable to determine a way to do this, so all three links are centered for now. Suggestions for ways to aesthetically separate or realign the links are welcome.

File: docs/walkthrough/redshift-estimator.rst [hunk 1]
```diff
@@ -0,0 +1,470 @@
...
+
+This operation will require two inputs (images and labels) and a neural network architecture. Finally, the operation produces one output, which is the trained classifier model. After all inputs, outputs, and attributes have been added, the structure of the operation should appear similar to the following:
+
+.. figure:: images/redshift-train-io.png
+    :align: center
+    :scale: 50%
+
+The code for this operation follows the standard procedure for creating and training a Keras network with one minor caveat. Keras models expect the output of a network during training to be an array or an array of indices. Because the outputs used in this tutorial are floating point values, they must be converted into one of these forms. This is the purpose of the *to_categorical* function below. The code for this operation is shown below.
```
Comment (brollb) [inline docs/walkthrough/redshift-estimator.rst:69]: > Keras models expect the output of a network during training to be an array or an array of indices. Because the outputs used in this tutorial are floating point values, they must be converted into one of these forms. This is the purpose of the to_categorical function below. The code for this operation is shown below.

My understanding was the biggest motivation for this function was due to changing this from a regression problem to a classification problem (ie, ensuring the output of the network was a probability distribution).

File: docs/walkthrough/redshift-estimator.rst [hunk 2]
```diff
@@ -0,0 +1,470 @@
...
+
+Note that the architecture selected from within the pipeline editor until after the `Neural Network Architecture`_ section of this guide is completed.
+
+.. figure:: images/redshift-t.png
+    :align: center
+    :scale: 50%
+
+Neural Network Architecture
+===========================
+This section will describe how to create a Convolutional Neural Network for estimating redshift from images. In particular, this section gives instructions on creating an `Inception-v1 network <https://towardsdatascience.com/a-simple-guide-to-the-versions-of-the-inception-network-7fc52b863202#8fff>`_. The basic structure of this network is an input block, a series of five inception blocks, followed by a densely connected classifier block. These blocks are each described in order below.
```
Comment (brollb) [inline docs/walkthrough/redshift-estimator.rst:139]: "Convolutional Neural Network" -> "convolution neural network"

File: docs/walkthrough/redshift-estimator.rst [hunk 3]
```diff
@@ -0,0 +1,470 @@
...
+
+For reference during design, the full architecture is shown below.
```
Comment (brollb) [inline docs/walkthrough/redshift-estimator.rst:141]: The image is too small to be able to see. Can you right click the architecture editor's background, expand all nodes, then download as PNG? Then you can probably link to it directly in case anyone wants to view it.

File: docs/walkthrough/redshift-estimator.rst [hunk 4]
```diff
@@ -0,0 +1,470 @@
...
+
+PredictRedshift Operation
+=========================
+This operation uses the model created by **TrainRedshift** to predict the values of a set on input images. This operation has no attributes, takes a model and a set of images as input and produces a set of predicted values (named *labels*) and the associates probability density functions that resulted in those values (named *pdfs*). The structure of the operation is as shown below:
+
+.. figure:: images/redshift-predict-io.png
+    :align: center
+    :scale: 50%
+   
+The *model.predict* function results in a pdf over all redshift values in the allowed range [0,0.4]. In order to get scalar values for predictions, a weighted average is taken for each pdf where the value being averaged is the redshift value represented by that bin and the weight is the pdf value at that bin (i.e. how likely it is that the value represented by that bin is the actual redshift value).
```
Comment (brollb) [inline docs/walkthrough/redshift-estimator.rst:220]: Probably should capitalize pdf -> PDF

**Review by brollb — CHANGES_REQUESTED** [review]
Nice work. Here are my comments:
- It doesn't look like the walkthrough has been added to the table of contents. I would add a new section before tutorials called "Step-by-Step Guides" and then add 3 sections - the introduction, CIFAR10, and Redshift Estimation.
- The index links don't seem to work for me.
- It might be worth using firefox for the screenshots since you can select HTML elements rather than manually trying to crop the images. Especially when it is something like a modal w/ rounded edges, this can be much easier.

File: docs/walkthrough/redshift-estimator.rst (hunk 1 above)
Comment (KyleAMoore) [reply docs/walkthrough/redshift-estimator.rst:69]: I believe I buried the lede on this section. That is the motivator for the function, but I was attempting to describe why the original data is not immediately compatible with classification training. I agree that the original wording is confusing and distracting. I have changed it to the below

> The code for this operation follows the standard procedure for creating and training a Keras network with one minor caveat. The method used by Pasquet et al. on which this pipeline is based formulates redshift prediction as a classification problem. Because the labels used in this tutorial are floating point values, they must be converted into a categorical format. This is the purpose of the *to_categorical* function. The code for this operation is shown below.

File: docs/walkthrough/executing-pipelines.rst [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+^^^^^^^^^^^^^^^
+In this section, you will select from the available compute backends. In the examples shown, for instance, the SciServer Compute service will be used. Each compute backend may require additional information, such as login credentials or computation resources that should be used.
+
+.. figure:: images/cifar-execute-compute.png
+    :align: center
+    :scale: 50%
+
+Storage Options
+^^^^^^^^^^^^^^^
+Here, the storage backend must be chosen from the available options. As with the compute options, SciServer's Files Service is used here as an example. Each backend may require additional input, such as login credentials and the desired storage location. This storage backend and location will be where all files created during execution will be stored. This will include both files used during execution, such as the generated python scripts, as well as artifacts created using Output operations.
```
[the remark this answers is not in the document]
Comment (KyleAMoore) [reply docs/walkthrough/executing-pipelines.rst:52]: That makes sense. Would it be correct if I changed "such as the generated python scripts," -> "such as data passed between operations,"? If not, I would just remove the clause entirely.

File: docs/walkthrough/creating-pipelines.rst [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    :align: center
+    :scale: 50%
+
+After an operation has been added, the attributes of that operation can be changed by clicking on the operation and then clicking on the current value of that attribute.
+
+.. figure:: images/cifar-gt.png
+    :align: center
+    :scale: 50%
+
+Operation inputs and outputs are represented by blue circles that are visible after clicking on the operation. Blue circles on the top on the operation represent inputs, while circles on the bottom represent outputs. The red X circle can be clicked to remove an operation from the pipeline. This does not remove it from the set of available operations. The green **</>** circle will open the operation editor view.
```
[the remark this answers is not in the document]
Comment (KyleAMoore) [reply docs/walkthrough/creating-pipelines.rst:26]: You're probably correct. I did not know about that feature, however. I added a short sentence to introduce it.

**Comment by KyleAMoore** [comment]
> It might be worth using firefox for the screenshots since you can select HTML elements rather than manually trying to crop the images. Especially when it is something like a modal w/ rounded edges, this can be much easier.

This is good advice for future work like this, but does this mean that you want me to replace the current images with new versions obtained via firefox? This would not be a problem, but it may take a few days if so, as the number of images means that it will take a fair amount of time to replace most/all of them.

**Comment by KyleAMoore** [comment]
> The index links don't seem to work for me.

Are you referring to the links on the `walkthrough.rst` page? When I tested them on the github rendering (https://github.com/KyleAMoore/deepforge/blob/1895-walkhrough/docs/walkthrough/walkthrough.rst), the links all worked in my testing. I am unsure of why they would not be working in whatever rendering engine you are using for testing.

File: docs/walkthrough/creating-neural-networks.rst [hunk 7]
```diff
@@ -0,0 +1,135 @@
+.. class:: centered
+
+    `<- Previous Tutorial <creating-operations.rst>`_
```
[the remark this answers is not in the document]
Comment (KyleAMoore) [reply docs/walkthrough/creating-neural-networks.rst:3]: I had excessive difficulty getting custom styles applied to the buttons. I found in the end that it works better overall to add all of the pages to the index. This lets shinx automatically create next and previous buttons at the bottom of every page.

File: docs/walkthrough/creating-neural-networks.rst [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    `Walkthrough Main Page <walkthrough.rst>`_
+    `Next Tutorial -> <executing-pipelines.rst>`_
+
+Creating Neural Networks
+------------------------
+
+Sections
```
[the remark this answers is not in the document]
Comment (KyleAMoore) [reply docs/walkthrough/creating-neural-networks.rst]: This should be fixed now. All walkthrough subpages have been added to the toctree. This means that they should all be visible and navigable from the sidebar. The `Sections` sections have been removed from each page.

**Review by brollb — CHANGES_REQUESTED** [review]
A couple minor changes then it should be all set!

**Review by brollb — APPROVED** [review]
